sopel/modules/ip.py

```diff
@@ -10,52 +10,242 @@
 
 from __future__ import unicode_literals, absolute_import, print_function, division
 
+import ipaddress
 import logging
 import os
 import socket
 import tarfile
+import typing
 
 import geoip2.database
-
-from sopel.config.types import FilenameAttribute, StaticSection
-from sopel.module import commands, example
-from sopel.tools import web
-
-urlretrieve = None
-try:
-    from urllib import urlretrieve
-except ImportError:
-    try:
-        # urlretrieve has been put under urllib.request in Python 3.
-        # It's also deprecated so this should probably be replaced with
-        # urllib2.
-        from urllib.request import urlretrieve
-    except ImportError:
-        pass
-
+import re
+import threading
+import urllib.request, json
+
+import sqlalchemy.sql
+
+from random import randint
+from urllib.request import urlretrieve
+
+#from minfraud import Client
+
+from sopel import module
+from sopel.config.types import FilenameAttribute, StaticSection, ValidatedAttribute, ListAttribute
+from sopel.tools import web, events, target, Identifier
+
+from sqlalchemy import Column, Integer, String, Float, Boolean, TIMESTAMP, Text
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.ext.declarative import declarative_base
+
+IRCCLOUD_IP = [
+    "2001:67c:2f08::/48",
+    "2a03:5180:f::/64",
+    "5.254.36.56/29",
+    "192.184.8.73",
+    "192.184.8.103",
+    "192.184.9.108",
+    "192.184.9.110",
+    "192.184.9.112",
+    "192.184.10.9",
+    "192.184.10.118",
+    ]
+
+IRCCLOUD_REASON = "IRCCloud"
+
+MIBBIT_IP = [
+    "207.192.75.252",
+    "64.62.228.82",
+    "78.129.202.38",
+    "109.169.29.95"
+    ]
+
+MIBBIT_REASON = "Mibbit"
+
+#Hardcoded for safety
+EXEMPT_IP = [
+    ("104.248.43.234", "Lazar, DigitalOcean"),
+    ("13.59.180.136", "War_ Limnoria bot"),
+    ("138.68.23.34", "Approved bot (Idlebot, aismallard)"),
+    ("18.132.171.104", "TARS"),
+    ("3.136.223.150", "bluesoul"),
+    ("54.174.11.206", "Helen"),
+    ("69.115.75.7", "Kufat and bots"),
+    ("94.159.196.226", "docazra, longtime user, IP is on dnsbl"),
+    ("2604:a880:2:d0::250:9001", "Approved bot (Idlebot, aismallard)"),
+    ("91.132.86.177", "bluesoul's bouncer"),
+    ("2001:470:1f07:13b::/64", "kufat's tunnelbroker"),
+    ]
 
 LOGGER = logging.getLogger(__name__)
-
+who_reqs = {}  # Keeps track of reqs coming from this plugin, rather than others
+
+# This dict will mostly be used as a list (walking each element) but having it as a dict
+# is useful for preventing duplicate entries.
+exemptions = {ipaddress.ip_network(i):reason for i, reason in EXEMPT_IP}
+exemptions.update(( (ipaddress.ip_network(i), IRCCLOUD_REASON) for i in IRCCLOUD_IP))
+exemptions.update(( (ipaddress.ip_network(i), MIBBIT_REASON) for i in MIBBIT_IP))
+
+BASE = declarative_base()
+
+safe_mode = True
```

Comment:
Having multiple `safe_mode` variables between different modules that could potentially go out-of-sync with each other has me a bit nervous. I feel like some kind of imported configuration object with the boolean as a field, shared between all, would be safer.

Reply:
Alternatively, in the bot instance where actions would otherwise be sent, it does the check _there_ and logs/actually runs accordingly

Reply:
Hmm, okay. I'm not sure that's very clear in the code, maybe name them different things (`email_safe_mode`, `ip_safe_mode`)